### ISSUE TITLE
Batch compiler is not hermetic in tests when compiler crashes

### ISSUE THREAD

**Issue by nshahan** [dart-lang/sdk#50044]
I've observed on the DDC test bots that sometimes a compiler crash in one test can cause failures in other tests that run after and I was able to construct a reproduction. I created a new test suite to demonstrate how one test crashing can cause another to fail.

1) Checkout https://dart-review.googlesource.com/c/sdk/+/260920. This includes changes to DDC that force one test to crash and trigger the error in the other test.
2) Build the dartdevc_test target: 
    `$ ./tools/build.py -m release dartdevc_test`.
3) Run the repro test suite and set the number of parallel tasks to 1. This will force the tests to run on the same batch mode task:
    `$ ./tools/test.py -r chrome -c dartdevk -j 1 tests/language/batch_mode_crash_repro`
4) Observe that both tests fail:
    `[00:04 | 100% | +    0 | -    2]`
5) Rename `a_crashes_compiler_test.dart` -> `c_crashes_compiler_test.dart`. The name of the test files determines the order that they run in so `b_should_pass_test.dart` will now run before `c_crashes_compiler_test.dart`
6) Run the test repro suite again:
    `$ ./tools/test.py -r chrome -c dartdevk -j 1 tests/language/batch_mode_crash_repro`
7) Observe that now one test passes and one test fails.
    `[00:06 | 100% | +    1 | -    1]`

cc @johnniwinther

**Comment by johnniwinther** [comment]
cc @jensjoha

**Comment by jensjoha** [comment]
I had to rename for both to crash...

The problem is this:

https://github.com/dart-lang/sdk/blob/2932261f835b32683b04cc70ccb7c77e8c3b7708/pkg/dev_compiler/lib/src/kernel/future_or_normalizer.dart#L29-L34

The `_FutureOrNormalizer` cache the old `FutureOr` in a static variable. It is thus cached with the first loaded platforms CoreTypes and by extension the first loaded platforms - in this case - `dart:async` and `Future` class inside it. That's then used to ask the class hierarchy about information, but the class hierarchy doesn't know about the library and throws.

Long story short: Don't save kernel libraries, CoreTypes instances etc in static variables.

**Comment by nshahan** [comment]
@jensjoha Thanks for pointing me in the right direction!
